lib/lte_uart_dfu/lte_uart_dfu.c

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2022 INTELLINIUM <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <drivers/uart.h>
+#include <string.h>
+#include <lte_uart_dfu.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include <dfu/dfu_target.h>
+#include <zephyr/logging/log.h>
+#include <stdnoreturn.h>
+#include <dfu_target_uart_encode.h>
+#include <dfu_target_uart_decode.h>
+#include <zephyr/sys/ring_buffer.h>
+
+LOG_MODULE_REGISTER(lte_uart_dfu, CONFIG_LTE_UART_DFU_LOG_LEVEL);
+
+K_THREAD_STACK_DEFINE(uart_handler_stack,
+		      CONFIG_LTE_UART_DFU_HANDLER_STACK_SIZE);
+static struct k_thread uart_handler_thread;
+static k_tid_t tid;
+static uint8_t in_buf[CONFIG_LTE_UART_DFU_IN_BUFFER_SIZE];
+
+struct lte_uart_dfu_ctx {
+	/* The UART device used for the DFU */
+	const struct device *uart_dev;
+
+	/* The ring buffer used to store received from the UART bus */
+	struct ring_buf rx_rb;
+
+	/* Semaphore used to wait data from UART bus */
+	struct k_sem lte_uart_sem;
+
+	/* Flag to check if the reception of application update has started */
+	bool download_started;
```

Comment:
Flag seems pointless: it is check only for a purpose of setting it, but no other logic depends on it.